scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+  node_vec = matrix(queue[,1], rows=1, cols=nrow(queue)) # reshape to force the creation of a new object
+  node_vec = matrix(node_vec, rows=nrow(queue), cols=1)  # reshape to force the creation of a new object
+  len = ncol(queue)
+  if (len < 2) {
+    new_queue = matrix(0,0,0)
+  } else {
+    new_queue = matrix(queue[,2:ncol(queue)], rows=nrow(queue), cols=ncol(queue)-1)
+  }
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    left: a 1xn matrix
+# INPUT:    right: a 1xn matrix
+# INPUT:    queue: a nxn matrix (queue containing values)
+# OUTPUT:   new_queue: the new queue nxn matrix
+dataQueuePush = function(Matrix[Double] left, Matrix[Double] right, Matrix[Double] queue)  return (Matrix[Double] new_queue) {
+  len = ncol(queue)
+  if(len <= 0) {
+    new_queue = cbind(left, right)
+  } else {
+    new_queue = cbind(queue, left, right)
+  }
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    X: a nxn matrix (the current samples with all features we observe)
+# INPUT:    y: a 1xn matrix (the current y to all observed samples)
+# OUTPUT:   lowest_residuals_index: the feature index with the lowest residuals
+findBestFeature = function(Matrix[Double] X, Matrix[Double] y, Integer sml_type) return (Integer lowest_residuals_index) {
+
+  lowest_residuals = 0
+  lowest_residuals_index = 1
+
+  for(i in 1: ncol(X)) {
+    current_feature = X[,i]
+
+    if(sml_type == 1) # Regression
+    {
+      weights = glm(X=current_feature, Y=y, dfam=1, verbose=FALSE)
+    }
+    else # Classification
+    {
+      assert(sml_type == 2)
+      weights = glm(X=current_feature, Y=y, dfam=2, verbose=FALSE)
+    }
+    y_residual = y - current_feature %*% weights
+    res = sum(y_residual ^ 2) # sum of least square calculation
+
+    if(i == 1 | res < lowest_residuals)
+    {
+      lowest_residuals = res
+      lowest_residuals_index = i
+    }
+  }
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    one_featureX: a 1xn matrix (one feature with all values)
+# OUTPUT:   best_split: the best split (highest gain indicates best splitting of datasets)
+findBestSplit = function(Integer sml_type, Matrix[Double] one_featureX, Double sim_score_parent, Matrix[Double] predictions, Double lambda)
+  return (Double best_split, Double best_gain)
+{
+  assert(ncol(one_featureX) == 1)
+  best_split = 0
+  best_gain = 0
+  best_sim_score_left = 0
+  best_sim_score_right = 0
+
+  ordered_X = orderFeature(one_featureX) # order entries by feature value
+
+  # iterate over all averages between 2 points
+  for(i in 1: (nrow(ordered_X)-1)) {
+    current_split = average(ordered_X[i,], ordered_X[i+1,])
+    [left, right] = splitMatrixByValue(one_featureX, current_split)
```

Comment:
Here you already know the index to split on, so use that to do the split not the value, (because you have to find that again)